designs/disruption-controls-by-method.md

```diff
@@ -0,0 +1,397 @@
+      // This cannot be of type intstr.IntOrString since kubebuilder doesn't support pattern
+      // checking for int nodes for IntOrString nodes.
+      // Ref: https://github.com/kubernetes-sigs/controller-tools/blob/55efe4be40394a288216dab63156b0a64fb82929/pkg/crd/markers/validation.go#L379-L388
+      // +kubebuilder:validation:Pattern:="^((100|[0-9]{1,2})%|[0-9]+)$"
+      // +kubebuilder:default:="10%"
+      Nodes string `json:"nodes" hash:"ignore"`
+      // Schedule specifies when a budget begins being active, following
+      // the upstream cronjob syntax. If omitted, the budget is always active.
+      // Timezones are not supported.
+      // This field is required if Duration is set.
+      // +kubebuilder:validation:Pattern:=`^(@(annually|yearly|monthly|weekly|daily|midnight|hourly))|((.+)\s(.+)\s(.+)\s(.+)\s(.+))$`
+      // +optional
+      Schedule *string `json:"schedule,omitempty" hash:"ignore"`
+      // Duration determines how long a Budget is active since each Schedule hit.
+      // Only minutes and hours are accepted, as cron does not work in seconds.
+      // If omitted, the budget is always active.
+      // This is required if Schedule is set.
+      // This regex has an optional 0s at the end since the duration.String() always adds
+      // a 0s at the end.
+      // +kubebuilder:validation:Pattern=`^([0-9]+(m|h)+(0s)?)$`
+      // +kubebuilder:validation:Type="string"
+      // +optional
+      Duration *metav1.Duration `json:"duration,omitempty" hash:"ignore"`
+}
+
+type DisruptionMethod string 
+
+const (
+	All DisruptionMethod = ""
+	Consolidation DisruptionMethod = "Consolidation" 
+	Drift DisruptionMethod = "Drift" 
+	Expiration DisruptionMethod = "Expiration" 
+	Emptiness DisruptionMethod = "Emptiness"
+)
+```
+
+
+##### Example
+```yaml
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+metadata:
+  name: default
+spec: # This is not a complete NodePool Spec.
+  disruption:
+    budgets:
+    - schedule: "* * * * *"
+      action: Drift
+      nodes: 10
+    - schedule: "* * * * *"
+      action: Drift
+      nodes: 10
+    # For all other actions , only allow 5 nodes to be disrupted at a time
+    - nodes: 5
+      schedule: "* * * * *"
+
+```
+
+In the original proposed spec, karpenter allows the user to specify up to [50 budgets](https://github.com/kubernetes-sigs/karpenter/blob/main/pkg/apis/v1beta1/nodepool.go#L96)
+
+If there are multiple active budgets, karpenter takes the most restrictive budget. This same principle will be applied to the disruption budgets in this approach. The only difference in behavior is that each window will either apply to a single action or all actions. 
+
+### Pros + Cons 
+* 👍 No nested definitions required 
+* 👍👍 Extending existing budgets api. No Breaking API Changes, completely backwards compatible  
+* 👎 With action being clearly tied to budgets, and other api logic being driven by disruption Method, we lose the chance to generalize per method controls 
```

Comment:
The second approach moves the `DisruptionMethod` to the top level of disruption.  The thing here is that there are probably going to be per action controls other than budgets(DoNotDisrupt by x action for example). 

So it might make sense to have some controls driven at a top level by action, as proposed in the second approach

Reply:
I see, you're thinking there might be some additional controls on a pod level eviction that you'd want to control based off what the disruption method is. I definitely agree that there are people who might want this in the future, but at the moment, I think it seems like another topic that I think we'll have a harder time coming to a consensus.

Reply:
yeah or controls at a node level. DisruptAfter x seconds, using method y.

I came to a similar conclusion that rather than designing housing for controls that don't exist today to just keep it simple.